### PR TITLE
jemalloc: Fix rebuild on change, fix performance regression

### LINKED ISSUE #1677: Hyrise singleton instance
@mrzzzrm and I just had a discussion about global state. The problems are
1. No one knows which variables hold state, as things like the singleton PluginManager are easily overseen
2. As a result, properly resetting the globally state in tests is difficult
3. The Static Initialization Order Fiasco

Having an instance that is constantly passed around bloats the interfaces unnecessarily. Also, we will likely never have more than one instance.

How about a single `Hyrise` class that holds the entire global state? Instead of `StorageManager::get()`, we would use `Hyrise::get().storage_manager`. Also, we could easily reset everything using `get() = Hyrise{}`.

### LINKED ISSUE #1925: jemalloc is not recompiled when the submodule is changed / fix huge_threshold
https://github.com/jemalloc/jemalloc/issues/1677

I have the fix ready, just adding this so that it doesn't get lost.

### PR DESCRIPTION
**Pull request by mrks**
fixes #1925

### CODE REVIEW

File: src/lib/hyrise.cpp [hunk 1]
```diff
@@ -1,8 +1,16 @@
 #include "hyrise.hpp"
 
+#include <jemalloc/jemalloc.h>
+
 namespace opossum {
 
 Hyrise::Hyrise() {
+  // Work around for a performance regression in jemalloc:
```
Comment (tobodner) [inline src/lib/hyrise.cpp]: *Workaround

File: src/lib/hyrise.cpp [hunk 2]
```diff
@@ -1,8 +1,16 @@
...
+  // https://github.com/jemalloc/jemalloc/issues/1677
+  // Other jemalloc tuning options can be added here as found necessary.
+  auto huge_threshold = 0;
```
Comment (tobodner) [inline src/lib/hyrise.cpp]: Did we play around with this value? In the referenced issue, the jemalloc developer states that 40M may be an appropriate value for the benchmark we showed them.

File: src/lib/hyrise.cpp (hunk 1 above)
Comment (mrks) [reply src/lib/hyrise.cpp]: Still WIP, no point in reviewing this yet. This part will have to go, anyway.

File: src/lib/hyrise.cpp (hunk 2 above)
Comment (mrks) [reply src/lib/hyrise.cpp]: Nope. For now, we don't really care about anyone else on the same machine, so giving back memory is not really an issue. I don't really want to open that box in addition to all the other open issues.

**Comment by mrks** [comment]
Unfortunately, that config option is read-only at runtime. As far as I understand jemalloc options, we need to set it before the binary starts. I found no way of setting it during compile time, so I was left with two options:

* Modify the jemalloc header file during configuration time using sed
* fork jemalloc

After trying the first, I found the second to be slightly (!) less ugly

**Comment by mrks** [comment]
Never mind - the old jemalloc was better anyway. I'll switch back to an old commit in the jemalloc repo.

**Comment by mrks** [comment]
Now it's ready for review. I can't check it on my Mac right now, as all my repos are quite dirty. Could you please verify that switching between jemalloc commits rebuilds jemalloc for Hyrise?

**Comment by tobodner** [comment]
> Never mind - the old jemalloc was better anyway. I'll switch back to an old commit in the jemalloc repo.

I too think that this is the cleanest way for now. Can you put a comment somewhere to remind us on the next try to bump the jemalloc version that we need to figure out a way to deal with huge_threshold?

**Comment by tobodner** [comment]
Will give this a close look tomorrow.

**Comment by mrks** [comment]
> I too think that this is the cleanest way for now. Can you put a comment somewhere to remind us on the next try to bump the jemalloc version that we need to figure out a way to deal with huge_threshold?

Not sure where/how to do that. In the end, either the new version is better or it is not. The whole problem this time came because when we looked at it, it was not rebuilt.

**Comment by Bouncner** [comment]
Still, without the comment, the guy who does a well-intended update of all submodules won’t know.

Can we assert the version via CMake?

**Comment by mrks** [comment]
> Still, without the comment, the guy who does a well-intended update of all submodules <s>won’t know.</s> without running the benchmarks should be punished.

I maintain that this was a one time issue caused by jemalloc not being rebuilt.

> Can we assert the version via CMake?

That would be the first and only place where actively guard code from being changed. As everywhere, if you touch something that could have performance implications (or you are unsure), you should run the numbers.

**Pull request merged**